### PR TITLE
refactor: convert LogViewer to React Virtuoso [WEB-1954]

### PR DESCRIPTION
**Pull request by EmilyBonar**
Convert LogViewer to use React-Virtuoso instead of react-window. [Here](https://github.com/determined-ai/hew/pull/103) is the PR performing the same conversion on the ChartGrid component for reference. There is a companion PR in the determined repo here: https://github.com/determined-ai/determined/pull/8862.

### Test Plan

Make sure there aren’t any feature regressions after the conversion, including but not limited to:
- Tailing
- Scroll to oldest
- Download logs
- Fullscreen
- Filter logs
- Fetch logs on scroll

### CODE REVIEW

File: src/kit/Section.tsx [hunk 1]
```diff
@@ -27,7 +34,9 @@ const Section: React.FC<Props> = ({ children, titleDivider = false, title }) =>
           <h5 className={titleClasses.join(' ')}>{title}</h5>
         </div>
       )}
-      <div className={css.body}>{children}</div>
+      <div className={css.body} id="section-body">
```
Comment (ashtonG) [inline src/kit/Section.tsx:37]: do we need this `id` attribute?

Comment (EmilyBonar) [reply src/kit/Section.tsx:37]: Good catch, no. That got left over from some testing.

File: src/kit/LogViewer/LogViewer.tsx [hunk 2]
```diff
@@ -350,30 +267,44 @@ const LogViewer: React.FC<Props> = ({
...
+      setTimeout(() => {
+        virtuosoRef.current?.scrollToIndex({
+          index: fetchDirection === FetchDirection.Older ? 'LAST' : 0,
+        });
         local.current.isScrollReady = true;
-      }
+      }, 200);
     });
-  }, [addLogs, canceler, fetchDirection, fetchLogs]);
+    // eslint-disable-next-line react-hooks/exhaustive-deps
```
Comment (ashtonG) [inline src/kit/LogViewer/LogViewer.tsx:306]: why are we skipping dependencies here?

File: src/kit/LogViewer/LogViewer.tsx [hunk 3]
```diff
@@ -420,53 +337,39 @@ const LogViewer: React.FC<Props> = ({
...
-      const listParent = logsRef.current.firstElementChild;
-      const list = listParent?.firstElementChild;
-      const scrollHeight = list?.scrollHeight ?? 0;
-      const parentHeight = listParent?.clientHeight ?? 0;
-      return scrollHeight > parentHeight;
-    });
-  }, [logs.length, logsRef]);
+  const handleReachedTop = useCallback(
+    (atTop: boolean) => {
+      if (atTop) handleFetchMoreLogs('start');
```
Comment (ashtonG) [inline src/kit/LogViewer/LogViewer.tsx:369]: when does this trigger if we've prepended 2 million log lines to the virtual list?

Comment (EmilyBonar) [reply src/kit/LogViewer/LogViewer.tsx:369]: That function triggers when the user scrolls to the top of the currently loaded logs. There aren't actually 2 million (or billion) items added to the list, that number is only used as an index. My understanding is that internally each item in the list has to have a non-negative index in order to keep track of what should be displayed. For an appending list that's straightforward - just start at 0 and add 1 for each additional item. However, in the case of a prepending list it gets more complicated. In order to ensure that the index never becomes negative as new logs are prepended we offset our starting index by an absurdly large number. Here's the doc page where you can see it in action: https://virtuoso.dev/prepend-items/

File: src/kit/LogViewer/LogViewer.tsx (hunk 2 above)
Comment (EmilyBonar) [reply src/kit/LogViewer/LogViewer.tsx:306]: I wanted to limit the initial log fetch to only happen on mount and when the direction changed and was concerned that the useEffect would fire too often with the full dep array.

Comment (EmilyBonar) [reply src/kit/LogViewer/LogViewer.tsx:306]: I think doing that broke searching through, so I'll reverse it.

**Review by ashtonG — COMMENTED** [review]
can we back out the select width changes?

**Comment by EmilyBonar** [comment]
@ashtonG I made the changes to `LogViewerSelect` because I was seeing the following happen on smaller widths:

Screenshot 2024-03-05 at 12-30-52 Experiment 5786 - Determined
The buttons on the right there are being cut off by the edge of the window, and the search box has shrunk to the point of being unusable. Let me know if you have a better solution to this.

The changes to `Select` were because I saw the text inside the select box sometimes wrap and cause the height of the element to double, which didn't look good. I can't seem to repro that now though, so I can revert those changes if you'd like.

**Comment by ashtonG** [comment]
let's back out the `Select.module.scss` changes from this pr for now.

**Review by ashtonG — APPROVED** [review]

**Pull request merged**